### ISSUE TITLE
[WASM] Error: 'oneHot' not yet implemented or not found in the registry 

### ISSUE THREAD

**Issue by syonfox** [tensorflow/tfjs#3114]
Hey it looks like oneHot is not implemented in the wasm Backend yet.

```
backend.ts:665 Uncaught (in promise) Error: 'oneHot' not yet implemented or not found in the registry. Did you forget to import the kernel?
    at Fa (backend.ts:665)
    at BackendWasm.t.oneHot (backend.ts:586)
```

#### Versions
```
tfjs: "1.7.1"
tfjs-converter: "1.7.1"
tfjs-core: "1.7.1"
tfjs-data: "1.7.1"
tfjs-layers: "1.7.1"

tf.wasm.version_wasm
"1.7.1"
```

Browser: ff 68.6.0esr

**Comment by dsmilkov** [comment]
Hi @syonfox,

Can you share some info about the ML model that you need this for, which helps us better prioritize?

**Comment by syonfox** [comment]
Sorry about the slow response.  I was using onehot in my median filter function  I tried to lay it out in a simmaler style to existing tensor flow function hope this helps. 
 If you wish to look at a usage I used it in a project here.  This is by no means a minimal example since lots of other stuff is happening.   The relivante code is pasted for the most part. 
http://syon.ca/cmpt414/pcv/src/pcv.html
the code is all in the http://syon.ca/cmpt414/pcv/src/js/main.js and http://syon.ca/cmpt414/pcv/src/js/medianFilter.js


Filter where one hot is used
```
function medianFilter(x, filter, strides, pad) {
    //make Kernal
    //todo allow for filter as array or number
    let filterSize = filter ** 2;
    let locs = tf.range(0, filterSize, filterSize );
    //makes a bunc of arrays each one reprensentin one of the valuesin the median window ie 2x2 filter i in chanle and 4 out chanles
    let f = tf.oneHot(tf.range(0,filterSize,1, 'int32'), filterSize).reshape([filter, filter, 1, filterSize]);
    let y = tf.conv2d(x,f,strides,pad);
    let m_idx = Math.floor(filterSize/2)+1;

    let top = tf.topk(y, m_idx, true);
    let median = tf.slice(top.values, [0,0,m_idx-1], [-1,-1,1] );

    return median;
}
```

Example Usage of Function
```
        res = res.reshape([imSize.x, imSize.y, 1])
        res = medianFilter(res, 3, [1, 1], 'same');
```

**Comment by oukigen** [comment]
The same error. Below is the complete code running on Node.js. The code runs successfully without WASM.
`// const tf = require('@tensorflow/tfjs');
const assert = require('assert');
const fs = require('fs');
const https = require('https');
const util = require('util');
const zlib = require('zlib');

const tf = require('@tensorflow/tfjs-node');
require('@tensorflow/tfjs-backend-wasm');
const argparse = require('argparse');

tf.setBackend('wasm');
tf.ready().then(() => {
	/**
	 * ******************data START******************
	 */
	const readFile = util.promisify(fs.readFile);

	// MNIST data constants:
	const BASE_URL = 'https://storage.googleapis.com/cvdf-datasets/mnist/';
	const TRAIN_IMAGES_FILE = 'train-images-idx3-ubyte';
	const TRAIN_LABELS_FILE = 'train-labels-idx1-ubyte';
	const TEST_IMAGES_FILE = 't10k-images-idx3-ubyte';
	const TEST_LABELS_FILE = 't10k-labels-idx1-ubyte';
	const IMAGE_HEADER_MAGIC_NUM = 2051;
	const IMAGE_HEADER_BYTES = 16;
	const IMAGE_HEIGHT = 28;
	const IMAGE_WIDTH = 28;
	const IMAGE_FLAT_SIZE = IMAGE_HEIGHT * IMAGE_WIDTH;
	const LABEL_HEADER_MAGIC_NUM = 2049;
	const LABEL_HEADER_BYTES = 8;
	const LABEL_RECORD_BYTE = 1;
	const LABEL_FLAT_SIZE = 10;

	// Downloads a test file only once and returns the buffer for the file.
	async function fetchOnceAndSaveToDiskWithBuffer(filename) {
	  return new Promise(resolve => {
	    const url = `${BASE_URL}${filename}.gz`;
	    if (fs.existsSync(filename)) {
	      resolve(readFile(filename));
	      return;
	    }
	    const file = fs.createWriteStream(filename);
	    console.log(`  * Downloading from: ${url}`);
	    https.get(url, (response) => {
	      const unzip = zlib.createGunzip();
	      response.pipe(unzip).pipe(file);
	      unzip.on('end', () => {
	        resolve(readFile(filename));
	      });
	    });
	  });
	}

	function loadHeaderValues(buffer, headerLength) {
	  const headerValues = [];
	  for (let i = 0; i < headerLength / 4; i++) {
	    // Header data is stored in-order (aka big-endian)
	    headerValues[i] = buffer.readUInt32BE(i * 4);
	  }
	  return headerValues;
	}

	async function loadImages(filename) {
	  const buffer = await fetchOnceAndSaveToDiskWithBuffer(filename);

	  const headerBytes = IMAGE_HEADER_BYTES;
	  const recordBytes = IMAGE_HEIGHT * IMAGE_WIDTH;

	  const headerValues = loadHeaderValues(buffer, headerBytes);
	  assert.equal(headerValues[0], IMAGE_HEADER_MAGIC_NUM);
	  assert.equal(headerValues[2], IMAGE_HEIGHT);
	  assert.equal(headerValues[3], IMAGE_WIDTH);

	  const images = [];
	  let index = headerBytes;
	  while (index < buffer.byteLength) {
	    const array = new Float32Array(recordBytes);
	    for (let i = 0; i < recordBytes; i++) {
	      // Normalize the pixel values into the 0-1 interval, from
	      // the original 0-255 interval.
	      array[i] = buffer.readUInt8(index++) / 255;
	    }
	    images.push(array);
	  }

	  assert.equal(images.length, headerValues[1]);
	  return images;
	}

	async function loadLabels(filename) {
	  const buffer = await fetchOnceAndSaveToDiskWithBuffer(filename);

	  const headerBytes = LABEL_HEADER_BYTES;
	  const recordBytes = LABEL_RECORD_BYTE;

	  const headerValues = loadHeaderValues(buffer, headerBytes);
	  assert.equal(headerValues[0], LABEL_HEADER_MAGIC_NUM);

	  const labels = [];
	  let index = headerBytes;
	  while (index < buffer.byteLength) {
	    const array = new Int32Array(recordBytes);
	    for (let i = 0; i < recordBytes; i++) {
	      array[i] = buffer.readUInt8(index++);
	    }
	    labels.push(array);
	  }

	  assert.equal(labels.length, headerValues[1]);
	  return labels;
	}

	/** Helper class to handle loading training and test data. */
	class MnistDataset {
	  constructor() {
	    this.dataset = null;
	    this.trainSize = 0;
	    this.testSize = 0;
	    this.trainBatchIndex = 0;
	    this.testBatchIndex = 0;
	  }

	  /** Loads training and test data. */
	  async loadData() {
	    this.dataset = await Promise.all([
	      loadImages(TRAIN_IMAGES_FILE), loadLabels(TRAIN_LABELS_FILE),
	      loadImages(TEST_IMAGES_FILE), loadLabels(TEST_LABELS_FILE)
	    ]);
	    this.trainSize = this.dataset[0].length;
	    this.testSize = this.dataset[2].length;
	  }

	  getTrainData() {
	    return this.getData_(true);
	  }

	  getTestData() {
	    return this.getData_(false);
	  }

	  getData_(isTrainingData) {
	    let imagesIndex;
	    let labelsIndex;
	    if (isTrainingData) {
	      imagesIndex = 0;
	      labelsIndex = 1;
	    } else {
	      imagesIndex = 2;
	      labelsIndex = 3;
	    }
	    const size = this.dataset[imagesIndex].length;
	    tf.util.assert(
	        this.dataset[labelsIndex].length === size,
	        `Mismatch in the number of images (${size}) and ` +
	            `the number of labels (${this.dataset[labelsIndex].length})`);

	    // Only create one big array to hold batch of images.
	    const imagesShape = [size, IMAGE_HEIGHT, IMAGE_WIDTH, 1];
	    const images = new Float32Array(tf.util.sizeFromShape(imagesShape));
	    const labels = new Int32Array(tf.util.sizeFromShape([size, 1]));

	    let imageOffset = 0;
	    let labelOffset = 0;
	    for (let i = 0; i < size; ++i) {
	      images.set(this.dataset[imagesIndex][i], imageOffset);
	      labels.set(this.dataset[labelsIndex][i], labelOffset);
	      imageOffset += IMAGE_FLAT_SIZE;
	      labelOffset += 1;
	    }

	    return {
	      images: tf.tensor4d(images, imagesShape),
	      labels: tf.oneHot(tf.tensor1d(labels, 'int32'), LABEL_FLAT_SIZE).toFloat()
	    };
	  }
	}
	/**
	 * ******************data END******************
	 */

	/**
	 * ******************model START******************
	 */
	const model = tf.sequential();
	model.add(tf.layers.conv2d({
	  inputShape: [28, 28, 1],
	  filters: 32,
	  kernelSize: 3,
	  activation: 'relu',
	}));
	model.add(tf.layers.conv2d({
	  filters: 32,
	  kernelSize: 3,
	  activation: 'relu',
	}));
	model.add(tf.layers.maxPooling2d({poolSize: [2, 2]}));
	model.add(tf.layers.conv2d({
	  filters: 64,
	  kernelSize: 3,
	  activation: 'relu',
	}));
	model.add(tf.layers.conv2d({
	  filters: 64,
	  kernelSize: 3,
	  activation: 'relu',
	}));
	model.add(tf.layers.maxPooling2d({poolSize: [2, 2]}));
	model.add(tf.layers.flatten());
	model.add(tf.layers.dropout({rate: 0.25}));
	model.add(tf.layers.dense({units: 512, activation: 'relu'}));
	model.add(tf.layers.dropout({rate: 0.5}));
	model.add(tf.layers.dense({units: 10, activation: 'softmax'}));

	const optimizer = 'rmsprop';
	model.compile({
	  optimizer: optimizer,
	  loss: 'categoricalCrossentropy',
	  metrics: ['accuracy'],
	});
	/**
	 * ******************model END******************
	 */

	/**
	 * ******************main START******************
	 */
	const data = new MnistDataset();

	async function run(epochs, batchSize, modelSavePath) {
		// await tf.setBackend('wasm');
	  await data.loadData();

	  const {images: trainImages, labels: trainLabels} = data.getTrainData();
	  model.summary();

	  // await tfnode.setBackend('wasm');

	  let epochBeginTime;
	  let millisPerStep;
	  const validationSplit = 0.15;
	  const numTrainExamplesPerEpoch =
	      trainImages.shape[0] * (1 - validationSplit);
	  const numTrainBatchesPerEpoch =
	      Math.ceil(numTrainExamplesPerEpoch / batchSize);
	  await model.fit(trainImages, trainLabels, {
	    epochs,
	    batchSize,
	    validationSplit
	  });

	  const {images: testImages, labels: testLabels} = data.getTestData();
	  const evalOutput = model.evaluate(testImages, testLabels);

	  console.log(
	      `\nEvaluation result:\n` +
	      `  Loss = ${evalOutput[0].dataSync()[0].toFixed(3)}; `+
	      `Accuracy = ${evalOutput[1].dataSync()[0].toFixed(3)}`);

	  if (modelSavePath != null) {
	    await model.save(`file://${modelSavePath}`);
	    console.log(`Saved model to path: ${modelSavePath}`);
	  }
	}

	const parser = new argparse.ArgumentParser({
	  description: 'TensorFlow.js-Node MNIST Example.',
	  addHelp: true
	});
	parser.addArgument('--epochs', {
	  type: 'int',
	  defaultValue: 20,
	  help: 'Number of epochs to train the model for.'
	});
	parser.addArgument('--batch_size', {
	  type: 'int',
	  defaultValue: 128,
	  help: 'Batch size to be used during model training.'
	})
	parser.addArgument('--model_save_path', {
	  type: 'string',
	  help: 'Path to which the model will be saved after training.'
	});
	const args = parser.parseArgs();
	// tfnode.setBackend('wasm').then(() => {
	// 	run(args.epochs, args.batch_size, args.model_save_path);
	// });
	run(args.epochs, args.batch_size, args.model_save_path);

	/**
	 * ******************main END******************
	 */
});`
